commit df539aab5887ce26f2cc284241caddf56e381dc7
Author:     Zhihong Zhang <100308595+nvidianz@users.noreply.github.com>
AuthorDate: Fri Sep 23 15:41:32 2022 -0400
Commit:     GitHub <noreply@github.com>
CommitDate: Fri Sep 23 15:41:32 2022 -0400

    Fixes several problems found during authz testing [skip ci] (#914)
    
    * 1. Fixed clone_job, not copying all fields.
    2. Fixed the deploy error when one site rejects the job.
    3. Admin API errors are logged now, instead feeding it to FOBS.
    4. Fixed Workspace decomposer problem caused by name->site_name change.
    
    * Log error for aux_send
---
 nvflare/private/fed/server/job_runner.py    | 23 +++++++++++++++--------
 nvflare/private/fed/server/server_engine.py | 29 ++++++++++++++++++++---------
 2 files changed, 35 insertions(+), 17 deletions(-)

diff --git a/nvflare/private/fed/server/job_runner.py b/nvflare/private/fed/server/job_runner.py
index 681b1568..d320b28e 100644
--- a/nvflare/private/fed/server/job_runner.py
+++ b/nvflare/private/fed/server/job_runner.py
@@ -16,7 +16,7 @@ import json
 import shutil
 import threading
 import time
-from typing import List
+from typing import List, Tuple
 
 from nvflare.apis.client import Client
 from nvflare.apis.event_type import EventType
@@ -78,7 +78,7 @@ class JobRunner(FLComponent):
         message.set_header(RequestHeader.JOB_META, json.dumps(job.meta))
         return message
 
-    def _deploy_job(self, job: Job, sites: dict, fl_ctx: FLContext) -> str:
+    def _deploy_job(self, job: Job, sites: dict, fl_ctx: FLContext) -> Tuple[str, list]:
         """Deploy the application to the list of participants
 
         Args:
@@ -86,7 +86,7 @@ class JobRunner(FLComponent):
             sites: participating sites
             fl_ctx: FLContext
 
-        Returns:  job id
+        Returns:  job id, failed_clients
 
         """
         fl_ctx.remove_prop(FLContextKey.JOB_RUN_NUMBER)
@@ -151,6 +151,7 @@ class JobRunner(FLComponent):
                 )
 
         abort_job = False
+        failed_clients = []
         if client_deploy_requests:
             engine = fl_ctx.get_engine()
             admin_server = engine.server.admin_server
@@ -159,7 +160,6 @@ class JobRunner(FLComponent):
             )
 
             # check replies and see whether required clients are okay
-            failed_clients = []
             for client_token, reply in client_token_to_reply.items():
                 client_name = client_token_to_name[client_token]
                 if reply:
@@ -189,7 +189,7 @@ class JobRunner(FLComponent):
             raise RuntimeError("deploy failure", deploy_detail)
 
         self.fire_event(EventType.JOB_DEPLOYED, fl_ctx)
-        return run_number
+        return run_number, failed_clients
 
     def _start_run(self, job_id: str, job: Job, client_sites: dict, fl_ctx: FLContext):
         """Start the application
@@ -329,7 +329,7 @@ class JobRunner(FLComponent):
                             try:
                                 self.log_info(fl_ctx, f"Got the job:{ready_job.job_id} from the scheduler to run")
                                 fl_ctx.set_prop(FLContextKey.CURRENT_JOB_ID, ready_job.job_id)
-                                job_id = self._deploy_job(ready_job, sites, fl_ctx)
+                                job_id, failed_clients = self._deploy_job(ready_job, sites, fl_ctx)
                                 job_manager.set_status(ready_job.job_id, RunStatus.DISPATCHED, fl_ctx)
 
                                 deploy_detail = fl_ctx.get_prop(FLContextKey.JOB_DEPLOY_DETAIL)
@@ -338,10 +338,17 @@ class JobRunner(FLComponent):
                                         ready_job.job_id, {JobMetaKey.JOB_DEPLOY_DETAIL.value: deploy_detail}, fl_ctx
                                     )
 
+                                if failed_clients:
+                                    deployable_clients = {
+                                        k: v for k, v in client_sites.items() if k not in failed_clients
+                                    }
+                                else:
+                                    deployable_clients = client_sites
+
                                 self._start_run(
                                     job_id=job_id,
                                     job=ready_job,
-                                    client_sites=client_sites,
+                                    client_sites=deployable_clients,
                                     fl_ctx=fl_ctx,
                                 )
                                 self.running_jobs[job_id] = ready_job
@@ -391,7 +398,7 @@ class JobRunner(FLComponent):
                 del self.running_jobs[job_id]
                 self.fire_event(EventType.JOB_ABORTED, fl_ctx)
             else:
-                raise RuntimeError(f"Job run: {job_id} does not exist.")
+                raise RuntimeError(f"Job {job_id} is not running.")
 
     def stop_all_runs(self, fl_ctx: FLContext):
         engine = fl_ctx.get_engine()
diff --git a/nvflare/private/fed/server/server_engine.py b/nvflare/private/fed/server/server_engine.py
index cecc8b31..e4e45b6e 100644
--- a/nvflare/private/fed/server/server_engine.py
+++ b/nvflare/private/fed/server/server_engine.py
@@ -51,7 +51,7 @@ from nvflare.apis.utils.fl_context_utils import get_serializable_data
 from nvflare.apis.workspace import Workspace
 from nvflare.fuel.hci.zip_utils import zip_directory_to_bytes
 from nvflare.fuel.utils import fobs
-from nvflare.private.admin_defs import Message
+from nvflare.private.admin_defs import Message, MsgHeader
 from nvflare.private.defs import RequestHeader, TrainingTopic
 from nvflare.private.fed.server.server_json_config import ServerJsonConfigurator
 from nvflare.private.fed.utils.fed_utils import security_close
@@ -581,12 +581,18 @@ class ServerEngine(ServerEngineInternalSpec):
             client_name = self.get_client_name_from_token(r.client_token)
             if r.reply:
                 try:
-                    results[client_name] = fobs.loads(r.reply.body)
-                except BaseException:
+                    error_code = r.reply.get_header(MsgHeader.RETURN_CODE, ReturnCode.OK)
+                    if error_code != ReturnCode.OK:
+                        self.logger.error(f"Aux message send error: {error_code} from client: {client_name}")
+                        shareable = make_reply(ReturnCode.ERROR)
+                    else:
+                        shareable = fobs.loads(r.reply.body)
+                    results[client_name] = shareable
+                except BaseException as e:
                     results[client_name] = make_reply(ReturnCode.COMMUNICATION_ERROR)
                     self.logger.error(
                         f"Received unexpected reply from client: {client_name}, "
-                        f"message body:{r.reply.body} processing topic:{topic}"
+                        f"message body:{r.reply.body} processing topic:{topic} Error:{e}"
                     )
             else:
                 results[client_name] = None
@@ -726,11 +732,16 @@ class ServerEngine(ServerEngineInternalSpec):
         for r in replies:
             site_name = self.get_client_name_from_token(r.client_token)
             if r.reply:
-                resp = fobs.loads(r.reply.body)
-                result[site_name] = (
-                    resp.get_header(ShareableHeader.CHECK_RESOURCE_RESULT, False),
-                    resp.get_header(ShareableHeader.RESOURCE_RESERVE_TOKEN, ""),
-                )
+                error_code = r.reply.get_header(MsgHeader.RETURN_CODE, ReturnCode.OK)
+                if error_code != ReturnCode.OK:
+                    self.logger.error(f"Client reply error: {r.reply.body}")
+                    result[site_name] = (False, "")
+                else:
+                    resp = fobs.loads(r.reply.body)
+                    result[site_name] = (
+                        resp.get_header(ShareableHeader.CHECK_RESOURCE_RESULT, False),
+                        resp.get_header(ShareableHeader.RESOURCE_RESERVE_TOKEN, ""),
+                    )
             else:
                 result[site_name] = (False, "")
         return result
